jobrunner/agent/task_api.py

```diff
@@ -0,0 +1,44 @@
+from jobrunner.lib import database  # cheating!
+from jobrunner.models import Task as ControllerTask  # cheating!
+from jobrunner.schema import AgentTask, TaskStage
+
+
+def get_active_tasks() -> list[AgentTask]:
+    """Get a list of active tasks from the controller"""
+    # cheating for now - should be HTTP API call
+    return [
+        AgentTask(
+            id=t.id,
+            type=t.type,
+            definition=t.definition,
+            created_at=t.created_at,
+        )
+        for t in database.find_where(ControllerTask, active=True)
+    ]
+
+
+def update_controller(
+    task: AgentTask,
+    stage: TaskStage,
+    timestamp: int,
+    results: dict = None,
+):
+    """Update the controller with the current state of the task.
+
+    stage: the current stage of the task from the agent's perspective
+    timestamp: ns timestamp of current stage was acheived (for accurate telemetry)
+    results: optional dictionary of completed results of this task, expected to be immutable
```

Comment:
Ah, this is an interesting difference in how I was thinking about things. In my head, I didn't have a specially, distinguished final results dict. Instead, the agent just sends up JSON blobs from the task, at any point the task wants to emit them, the details of which are determined by the task itself. Then it's up the controller to decide that a task is done.

I do think we'll have a need for some kind of mutable data that's more than just the `stage` enum so that tasks can provide more complex information about what they're currently up to. Whether that's in addition to the `stage` or subsumes it, I'm less sure.

We should probably discuss all this directly rather than on the PR.

Reply:
So, some stuff will be the same for all tasks e.g. a stage value,a dn the timestamp in happened.

Whereas the results json format will be different per task type.

I get that we may want to send more that stage up as general task metadata, but I don't think we should mix that with task-specific results, I'd prefer to keep them separate.

We can add a completed boolean flag easily enough.